buy_together_app/views.py

```diff
@@ -7,3 +9,16 @@ def main_page(request):
 
 def description_page(request):
     return render(request, 'buy_together_app/description.html')
+
+
+def log_in_page(request):
```

Comment:
We need to have this tested as well. It's a bit more complex than other test cases you did up until now, as it requires calling some django methods. You can use monkeypatch to mock some of those methods and make the test suite use that mock. This way, you don't need to rely on everything being in place, and you can also verify that everything was called as you expected.

I'm going to provide a simplistic approach to go about it. It could be done in a much better way, but for now it should do.

The basics:

```
from unittest.mock import Mock

mock = Mock()
```
This creates a mock object that you can make it take and return whatever you want

```
monkeypatch.setattr(foo, foo.object.__name__, mock)
```
This makes the object called `mock` replace the object called `foo.object.__name__` in module `foo`.

More specifically, suppose that this is the module you're testing (called foo.py):

```
from django.shortcuts import render, redirect
from django.contrib.auth import authenticate, login
from django.contrib import messages


def log_in_page(request):
    if request.method == 'POST':
        user_name = request.POST.get('username')
        user_pass = request.POST.get('password')
        user = authenticate(request, username=user_name, password=user_pass)
        if user is not None:
            login(request, user)
            return redirect('Main Page')
        else:
            messages.info(request, 'User name OR Password is incorrect')
    return render(request, 'buy_together_app/login.html')
```

This is how your tests will look like:
```
import pytest
from unittest.mock import Mock
import foo


class Test_Login:
    @pytest.fixture
    def authenticate_mock(self, monkeypatch):
        mock = Mock(return_value=Mock())
        monkeypatch.setattr(foo, foo.authenticate.__name__, mock)
        return mock

    @pytest.fixture
    def login_mock(self, monkeypatch):
        mock = Mock()
        monkeypatch.setattr(foo, foo.login.__name__, mock)
        return mock

    @pytest.fixture
    def redirect_mock(self, monkeypatch):
        mock = Mock()
        monkeypatch.setattr(foo, foo.redirect.__name__, mock)
        return mock

    @pytest.fixture
    def messages_mock(self, monkeypatch):
        mock = Mock()
        monkeypatch.setattr(foo, "messages", mock)
        return mock

    @pytest.fixture
    def render_mock(self, monkeypatch):
        mock = Mock()
        monkeypatch.setattr(foo, foo.render.__name__, mock)
        return mock

    def test_login_succuss(
        self, authenticate_mock, login_mock, redirect_mock, messages_mock, render_mock
    ):
        request_mock = Mock()
        request_mock.method = "POST"
        output = foo.log_in_page(request_mock)
        assert output == redirect_mock.return_value
        assert authenticate_mock.call_count == 1
        assert login_mock.call_count == 1
        assert redirect_mock.call_count == 1
        assert messages_mock.info.call_count == 0
        assert render_mock.call_count == 0

    def test_login_not_post(
        self, authenticate_mock, login_mock, redirect_mock, messages_mock, render_mock
    ):
        request_mock = Mock()
        request_mock.method = "NOT_POST"
        output = foo.log_in_page(request_mock)
        assert output == render_mock.return_value
        assert authenticate_mock.call_count == 0
        assert login_mock.call_count == 0
        assert redirect_mock.call_count == 0
        assert messages_mock.info.call_count == 0
        assert render_mock.call_count == 1

    def test_login_failed_auth(
        self, authenticate_mock, login_mock, redirect_mock, messages_mock, render_mock
    ):
        request_mock = Mock()
        request_mock.method = "POST"
        authenticate_mock.return_value = None
        output = foo.log_in_page(request_mock)
        assert output == render_mock.return_value
        assert authenticate_mock.call_count == 1
        assert login_mock.call_count == 0
        assert redirect_mock.call_count == 0
        assert messages_mock.info.call_count == 1
        assert render_mock.call_count == 1

```

Reply:
thanks! did saw some different ways to test the view functions, will use them next testing.
thou the code was REALLY helpful to understand how to test.

Reply:
Glad it helped!